### PR TITLE
perf(@desktop): external calls optimization when switching chats

### LINKED ISSUE #1642: Switching between chats on macOS is slow
~~when you just run the app, the delay is minimum, like 1s, but if you scroll and get more messages from history or receive from mailserver, it's geting growing and might increase up to 5s~~

when you click on any chat, app is freezing for a few seconds, its pretty general action, and when you surf between chats the app feels slow and laggy

### PR DESCRIPTION
**Pull request by pinebit**
Probably fixes #1642 

Here I tried to optimize the RPC calls (assuming they are expensive) being made when the user is switching chats:
- getContactByID
- markAllRead
- statusUpdated

### CODE REVIEW

File: src/status/contacts.nim [hunk 1]
```diff
@@ -17,41 +19,55 @@ type
...
 proc newContactModel*(events: EventEmitter): ContactModel =
     result = ContactModel()
+    result.contactsMap = initTable[string, Profile]()
     result.events = events
 
 proc getContactByID*(self: ContactModel, id: string): Profile =
+  if self.contactsMap.hasKey(id):
+    return self.contactsMap[id]
+
+  echo "*** getContactByID cache miss: ", id
```
Comment (richard-ramos) [inline src/status/contacts.nim:29]:
```suggestion
```

File: src/app/chat/event_handling.nim [hunk 2]
```diff
@@ -10,6 +10,8 @@ proc handleChatEvents(self: ChatController) =
   # Display already saved messages
   self.status.events.on("messagesLoaded") do(e:Args):
     let evArgs = MsgsLoadedArgs(e)
+    if evArgs.messages.len > 0:
+      self.status.chat.statusUpdates()
```
Comment (richard-ramos) [inline src/app/chat/event_handling.nim:14]: I'm thinking if it wouldnt be better if instead of having this code here inside on `messagesLoaded`, instead call that function inside `proc init*(self: ChatController)` in src/app/chat/core.nim, so it's only once when you login.

File: src/app/chat/views/channel.nim [hunk 3]
```diff
@@ -93,13 +93,15 @@ QtObject:
     if (self.chats.chats.len == 0): return
     let selectedChannel = self.getChannel(channelIndex)
     if (selectedChannel == nil): return
-    discard self.status.chat.markAllChannelMessagesRead(selectedChannel.id)
+    if (selectedChannel.unviewedMessagesCount > 0 or selectedChannel.mentionsCount > 0):
+      discard self.status.chat.markAllChannelMessagesRead(selectedChannel.id)
```
Comment (richard-ramos) [inline src/app/chat/views/channel.nim:97]: It might be a good idea to try to run this async, or pass exactly the message ids that are not read (there's a function for marking as seen an array of message ids in status-go), since `markAllChannelMessagesRead` is really slow

File: src/app/chat/views/channel.nim [hunk 4]
```diff
@@ -93,13 +93,15 @@ QtObject:
...
 
   proc markChatItemAsRead*(self: ChannelView, id: string) {.slot.} =
     if (self.chats.chats.len == 0): return
     let selectedChannel = self.getChannelById(id)
     if (selectedChannel == nil): return
-    discard self.status.chat.markAllChannelMessagesRead(selectedChannel.id)
+    if (selectedChannel.unviewedMessagesCount > 0 or selectedChannel.mentionsCount > 0):
+      discard self.status.chat.markAllChannelMessagesRead(selectedChannel.id)
```
Comment (richard-ramos) [inline src/app/chat/views/channel.nim:104]: Same observation as before

File: src/app/chat/views/channel.nim [hunk 5]
```diff
@@ -163,7 +165,8 @@ QtObject:
     let selectedChannel = self.getChannelById(channel)
     self.activeChannel.setChatItem(selectedChannel)
     
-    discard self.status.chat.markAllChannelMessagesRead(self.activeChannel.id)
+    if (selectedChannel.unviewedMessagesCount > 0 or selectedChannel.mentionsCount > 0):
+      discard self.status.chat.markAllChannelMessagesRead(self.activeChannel.id)
```
Comment (richard-ramos) [inline src/app/chat/views/channel.nim:169]: Same observation as before

File: src/status/libstatus/contacts.nim [hunk 6]
```diff
@@ -1,45 +1,34 @@
...
-var
-  contacts {.threadvar.}: JsonNode
-  contactsInited {.threadvar.}: bool
-  dirty: Atomic[bool]
+# this module is made stateless intentionally
+# all caching logic is done in status/contacts.nim
 
-# TODO: remove Profile from here
-proc blockContact*(contact: Profile): string =
+proc blockContact*(id: string, ensVerified: bool, ensName: string, alias: string, identicon: string, systemTags: seq[string], localNickname: string): string =
```
Comment (jrainville) [inline src/status/libstatus/contacts.nim:7]: Why did you change this to pass all individually instead? I don't think it should make anyway performance gain and we call this function using a Profile anyway.

File: src/status/contacts.nim [hunk 7]
```diff
@@ -9,6 +9,8 @@ const DELETE_CONTACT* = "__deleteThisContact__"
 
 type
   ContactModel* = ref object
+    # contact.ID => Profile
+    contactsMap: Table[string, Profile]
```
Comment (jrainville) [inline src/status/contacts.nim:13]: Is there a chance that the contactsMap could be accessed asynchronously? If so, I wonder if it could cause crashes.

Comment (pinebit) [reply src/status/contacts.nim:13]: That's why I need your review, since you know the codebase much better than me.  Who could confirm this code can be accessed asynchronously or not?

**Comment by iurimatias** [comment]
testing this PR again, it feels an improvement but it's still a massive difference seen when disabling the contacts & mark as read calls on master which cause channel switching to be instant

**Comment by iurimatias** [comment]
closing, fixed in different PR
